### PR TITLE
Fix AB 218 - Files with lot of unsummarized Ops. 

### PR DESCRIPTION
**Pull request by NicholasCouri**
The problem happens when we generate thousands of ops in a short period of time and get throttled by the push service. The users keep sending ops and, in case we are eventually able to recover, after retrying to connect (respecting imposed delay by the service) and submitting the outstanding ops, we reach a point in which the Summarizer needs to generate a snapshot as the server is now throttling the clients due to the limit of unsummarized ops. 

At this point, the server starts to send the following error: **NACK (ThrottlingError): Submit a summary before inserting additional ops.**

From this moment on, the client goes into in a loop trying to launch a summarizer and immediately asking it to stop as the parent of the summarizer is getting disconnected/throttled after trying to send an op (probably join ?). 

 
## Description

Before this change, if the parent is not connected, we would immediately ask the summarizer to stop.  With these changes, we will only ask the summarizer to stop in case the last summary would not be generated ( stopReason !=="parentNotConnected" )

Notice that we considered removing the entire short-circuit unconditionally but decided not to as to preserve resources as launching the summarizer against a scenario that is guaranteed to NOT generate the last summary would be a waste of 
resources. 

I took the chance to add the stop reason for the scenario in which we return early. 

## Does this introduce a breaking change?
No.

## Any relevant information

I did run the stress tests under the debugger and, when the condition we care about was hit, we successfully summarized. 

As a side not from the impact of these changes, in the last 30 days: 
 - 231638 documents hit the  "early exit after starting summarizer" 

Note: I still need to run the test coverage to see if it does not timeout.

### CODE REVIEW

**Comment by tyler-cai-microsoft** [comment]
Is there a way to replicate this in the tests?

**Comment by NicholasCouri** [comment]
> Is there a way to replicate this in the tests?

Just added it :)

**Review by tyler-cai-microsoft — APPROVED** [review]

**Review by kian-thompson — APPROVED** [review]

File: packages/runtime/container-runtime/src/summaryManager.ts [hunk 1]
```diff
@@ -213,11 +218,21 @@ export class SummaryManager implements IDisposable {
...
 
             // Re-validate that it need to be running. Due to asynchrony, it may be not the case anymore
+            // If we can't run the LastSummary, simply return as to avoid paying the cost of launching
+            // the summarizer at all.
             const shouldSummarizeState = this.getShouldSummarizeState();
             if (shouldSummarizeState.shouldSummarize === false) {
-                this.state = SummaryManagerState.Starting;
-                summarizer.stop(shouldSummarizeState.stopReason);
-                return "early exit after starting summarizer";
+                // In order to allow the last summary to run, we need not only a stop reason that would
```
Comment (jatgarg) [inline packages/runtime/container-runtime/src/summaryManager.ts:225]: nit: "need not only" -> "not only need"

File: packages/runtime/container-runtime/src/summaryManager.ts [hunk 2]
```diff
@@ -213,11 +218,21 @@ export class SummaryManager implements IDisposable {
...
+                // allow it but also, startWithDelay to be false (start the summarization immediately),
```
Comment (jatgarg) [inline packages/runtime/container-runtime/src/summaryManager.ts:226]: nit: "startWithDelay" -> "startWithInitialDelay"

**Review by jatgarg — APPROVED** [review]

**Pull request merged**